src/python_testing/TC_DGSW_2_3.py

```diff
@@ -0,0 +1,201 @@
+
+    @staticmethod
+    def is_valid_uint32_value(value):
+        return isinstance(value, int) and 0 <= value <= 0xFFFFFFFF
+
+    async def read_dgsw_attribute_expect_success(self, endpoint, attribute):
+        cluster = Clusters.Objects.SoftwareDiagnostics
+        return await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=attribute)
+
+    async def send_reset_watermarks_command(self):
+        endpoint = 0
+        diags_cluster = Clusters.Objects.SoftwareDiagnostics
+
+        try:
+            response = await self.send_single_cmd(cmd=diags_cluster.Commands.ResetWatermarks(), endpoint=endpoint)
+        except InteractionModelError as e:
+            asserts.assert_equal(e.status, Status.Success, "Unexpected error returned")
+            pass
+
+        return response
+
+    def desc_TC_DGSW_2_3(self) -> str:
+        """Returns a description of this test"""
+        return "[TC-DGSW-2.1] Attributes with Server as DUT"
+
+    def pics_TC_DGSW_2_3(self) -> list[str]:
+        return ["DGSW.S"]
+
+    def steps_TC_DGSW_2_3(self) -> list[TestStep]:
+        steps = [
+            TestStep(1, "Commissioning, already done", is_commissioning=True),
+            TestStep(2, "Read the AttributeList attribute"),
+            TestStep(3, "Read the ThreadMetrics attribute"),
+            TestStep(4, "Read the CurrentHeapHighWatermark attribute"),
+            TestStep(5, "Read the CurrentHeapUsed attribute"),
+            TestStep(6, "Send ResetWatermarks command"),
+            TestStep(7, "Read the CurrentHeapHighWatermark attribute"),
+            TestStep(8, "Read the ThreadMetrics attribute"),
+        ]
+        return steps
+
+    @async_test_body
+    async def test_TC_DGSW_2_3(self):
+
+        endpoint = self.get_endpoint(default=0)
+
+        # STEP 1: Commission DUT (already done)
+        self.step(1)
+
+        # STEP 2: TH reads from the DUT the AttributeList attribute
+        self.step(2)
+        attributes = Clusters.SoftwareDiagnostics.Attributes
+        attribute_list = await self.read_dgsw_attribute_expect_success(endpoint=endpoint, attribute=attributes.AttributeList)
+
+        # STEP 3: TH reads from the DUT the ThreadMetrics attribute
+        self.step(3)
+        if self.pics_guard(Clusters.SoftwareDiagnostics.Attributes.ThreadMetrics.attribute_id in attribute_list):
+            thread_metrics_original = await self.read_dgsw_attribute_expect_success(endpoint=endpoint, attribute=attributes.ThreadMetrics)
+            # the Id field is mandatory
+            asserts.assert_true(self.is_valid_uint64_value(thread_metrics_original[0].id), "Id field should be a uint64 type")
+            if thread_metrics_original[0].name is not None:
+                asserts.assert_true(thread_metrics_original[0].name, str, "Name field should be a string type")
+            if thread_metrics_original[0].stackFreeCurrent is not None:
+                asserts.assert_true(self.is_valid_uint32_value(
```

Comment:
please add a followup issue to have extra asserts as this is overly verbose (`asserts.assert_true(self.is_balid_uint32_value(...)` seems like it should read like `self.assert_valid_uint32(...)`)

Reply:
https://github.com/project-chip/connectedhomeip/issues/36968